File: cli/src/core/ProcessedSourceStore.test.ts
import { describe, expect, it } from "vitest";
import type { FileWrite } from "../Types.js";
import {
	addProcessed,
	carryProcessedSummaryHash,
	emptyProcessedSet,
	hasProcessed,
	readProcessedSet,
	saveProcessedSet,
} from "./ProcessedSourceStore.js";
import type { StorageProvider } from "./StorageProvider.js";
import type { SourceRef } from "./TopicKBTypes.js";

function makeFakeStorage(initial: Record<string, string> = {}): StorageProvider {
	const files = new Map(Object.entries(initial));
	return {
		readFile: async (p: string) => files.get(p) ?? null,
		writeFiles: async (fws: FileWrite[]) => {
			for (const f of fws) files.set(f.path, f.content);
		},
		listFiles: async (prefix: string) => [...files.keys()].filter((k) => k.startsWith(prefix)),
		exists: async () => true,
		ensure: async () => {},
	};
}

const ref = (type: SourceRef["type"], id: string): SourceRef => ({ type, id, timestamp: "2026-01-01T00:00:00Z" });

describe("ProcessedSourceStore", () => {
	it("returns an empty set when the file is absent", async () => {
		const set = await readProcessedSet("/tmp/x", makeFakeStorage());
		expect(set).toEqual(emptyProcessedSet());
	});

	it("hasProcessed reflects membership by type+id", () => {
		const set = addProcessed(emptyProcessedSet(), [ref("summary", "abc"), ref("plan", "p1")]);
		expect(hasProcessed(set, ref("summary", "abc"))).toBe(true);
		expect(hasProcessed(set, ref("plan", "p1"))).toBe(true);
		expect(hasProcessed(set, ref("summary", "p1"))).toBe(false); // same id, wrong type
		expect(hasProcessed(set, ref("note", "abc"))).toBe(false);
	});

	it("addProcessed is idempotent and immutable", () => {
		const base = emptyProcessedSet();
		const once = addProcessed(base, [ref("summary", "abc")]);
		const twice = addProcessed(once, [ref("summary", "abc")]);
		expect(twice.processed.summary).toEqual(["abc"]);
		expect(base.processed.summary).toEqual([]); // original untouched
	});

	it("round-trips through save/read", async () => {
		const storage = makeFakeStorage();
		const set = addProcessed(emptyProcessedSet(), [ref("userfile", "a.md@deadbeef")]);
		await saveProcessedSet(set, "/tmp/x", storage);
		const back = await readProcessedSet("/tmp/x", storage);
		expect(back).toEqual(set);
	});

	it("normalizes a partial on-disk shape to all four buckets", async () => {
		const storage = makeFakeStorage({
			"topics/processed.json": JSON.stringify({ schemaVersion: 1, processed: { summary: ["x"] } }),
		});
		const set = await readProcessedSet("/tmp/x", storage);
		expect(set.processed).toEqual({ summary: ["x"], plan: [], note: [], userfile: [] });
	});

	it("returns empty on unparseable JSON", async () => {
		const storage = makeFakeStorage({ "topics/processed.json": "{not json" });
		const set = await readProcessedSet("/tmp/x", storage);
		expect(set).toEqual(emptyProcessedSet());
	});

	it("normalizes valid JSON with no `processed` key to all four empty buckets", async () => {
		const storage = makeFakeStorage({ "topics/processed.json": JSON.stringify({ schemaVersion: 1 }) });
		const set = await readProcessedSet("/tmp/x", storage);
		expect(set).toEqual(emptyProcessedSet());
	});

	describe("carryProcessedSummaryHash (rebase 1:1 migration)", () => {
		it("carries the old hash's mark onto the new hash AND prunes the dead old hash", async () => {
			const storage = makeFakeStorage();
			await saveProcessedSet(addProcessed(emptyProcessedSet(), [ref("summary", "OLD")]), "/tmp/x", storage);
			await carryProcessedSummaryHash("/tmp/x", "OLD", "NEW", storage);
			const set = await readProcessedSet("/tmp/x", storage);
			expect(hasProcessed(set, ref("summary", "NEW"))).toBe(true);
			// The old commit no longer exists after a 1:1 rebase migration — its entry is
			// pruned so the set doesn't grow by one dead row per rebased commit.
			expect(hasProcessed(set, ref("summary", "OLD"))).toBe(false);
			expect(set.processed.summary).toEqual(["NEW"]);
		});

		it("does nothing when old and new hash are identical", async () => {
			const storage = makeFakeStorage();
			await saveProcessedSet(addProcessed(emptyProcessedSet(), [ref("summary", "SAME")]), "/tmp/x", storage);
			await carryProcessedSummaryHash("/tmp/x", "SAME", "SAME", storage);
			const set = await readProcessedSet("/tmp/x", storage);
			expect(set.processed.summary).toEqual(["SAME"]);
		});

		it("is a no-op when the old hash was never processed", async () => {
			const storage = makeFakeStorage();
			await carryProcessedSummaryHash("/tmp/x", "OLD", "NEW", storage);
			const set = await readProcessedSet("/tmp/x", storage);
			expect(hasProcessed(set, ref("summary", "NEW"))).toBe(false);
		});

		it("is idempotent when the new hash is already processed", async () => {
			const storage = makeFakeStorage();
			await saveProcessedSet(
				addProcessed(emptyProcessedSet(), [ref("summary", "OLD"), ref("summary", "NEW")]),
				"/tmp/x",
				storage,
			);
			await carryProcessedSummaryHash("/tmp/x", "OLD", "NEW", storage);
			const set = await readProcessedSet("/tmp/x", storage);
			expect(set.processed.summary).toEqual(["OLD", "NEW"]);
		});
	});
});

File: intellij/src/test/kotlin/ai/jolli/jollimemory/core/ProcessedSourceStoreTest.kt
package ai.jolli.jollimemory.core

import io.kotest.matchers.booleans.shouldBeFalse
import io.kotest.matchers.booleans.shouldBeTrue
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.nio.file.Path

class ProcessedSourceStoreTest {

    @TempDir
    lateinit var tempDir: Path
    private lateinit var storage: FolderStorage

    @BeforeEach
    fun setUp() {
        val kbRoot = tempDir.resolve("kb")
        storage = FolderStorage(kbRoot, MetadataManager(kbRoot.resolve(".jolli")))
        storage.ensure()
    }

    @Test
    fun `emptyProcessedSet serializes with all four buckets in canonical order`() {
        val expected = listOf(
            "{",
            "\t\"schemaVersion\": 1,",
            "\t\"processed\": {",
            "\t\t\"summary\": [],",
            "\t\t\"plan\": [],",
            "\t\t\"note\": [],",
            "\t\t\"userfile\": []",
            "\t}",
            "}",
        ).joinToString("\n")
        TopicJson.stringify(ProcessedSourceStore.emptyProcessedSet()) shouldBe expected
    }

    @Test
    fun `hasProcessed and addProcessed are idempotent`() {
        val ref = SourceRef(SourceType.SUMMARY, "hash1", "t")
        var set = ProcessedSourceStore.emptyProcessedSet()
        ProcessedSourceStore.hasProcessed(set, ref).shouldBeFalse()

        set = ProcessedSourceStore.addProcessed(set, listOf(ref))
        ProcessedSourceStore.hasProcessed(set, ref).shouldBeTrue()

        // Adding again does not duplicate.
        set = ProcessedSourceStore.addProcessed(set, listOf(ref))
        set.processed[SourceType.SUMMARY] shouldBe listOf("hash1")
    }

    @Test
    fun `save then read round-trips through the provider`() {
        val set = ProcessedSourceStore.addProcessed(
            ProcessedSourceStore.emptyProcessedSet(),
            listOf(
                SourceRef(SourceType.SUMMARY, "h1", "t"),
                SourceRef(SourceType.PLAN, "p1", "t"),
            ),
        )
        ProcessedSourceStore.saveProcessedSet(set, storage)
        val read = ProcessedSourceStore.readProcessedSet(storage)
        read.processed[SourceType.SUMMARY] shouldBe listOf("h1")
        read.processed[SourceType.PLAN] shouldBe listOf("p1")
        read.processed[SourceType.NOTE] shouldBe emptyList()
    }

    @Test
    fun `read of missing file yields the empty set`() {
        val read = ProcessedSourceStore.readProcessedSet(storage)
        read.processed[SourceType.USERFILE] shouldBe emptyList()
    }
}
